### DISCUSSION TITLE
Datetime xTickFormat in a jupyter widget

### DISCUSSION THREAD

**Discussion by frtennis1** [uwdata/mosaic#238] [Q&A]
Have had a great time using Mosaic so far - thank you for all the work building it!

I hit a snag maybe someone will know how to overcome. I can't seem to draw a plot with datetime-formatted xTicks using the Jupyter widget (i.e. json encoding). An MRE of what I'm trying to accomplish,

```
import numpy as np
import pandas as pd
from mosaic_widget import MosaicWidget

timestamps = pd.date_range('20230525-09:30', '20230530 10:00', freq='1min')
df = pd.DataFrame({'ts': timestamps, 'y': np.sin(np.arange(len(timestamps)) / 300)})

spec = {
    "plot": [
        {
            "mark": "lineY",
            "data": {"from": "df", "filterBy": "$filt"},
            "x": "ts",
            "y": "y",
        }
    ]
}

MosaicWidget(spec, data={'df': df})
```

which draws the plot except the x-ticks are formatted for integers instead of datetimes.

## Things I've tried

1. I can get this to work if I'm directly writing javascript by passing a function (e.g. from [d3-time-format](https://d3js.org/d3-time-format)) to `vg.xTickFormat`. Unfortunately, for my current use-case accomplishing this via the Jupyter widget is important to me.
2. I tried adding a format string to `{"xTickFormat": "<some string>"}` to the spec, but it seems that the only tick formats accepted are [d3-format](https://d3js.org/d3-format) strings which only operate on numbers and not datetimes.
3. I tried passing in a parameter `{"xTickFormat": "$myTimeFormatter"}` but there didn't seem a way to add a function paremeter via the json spec, and defining `$myTimeFormatter = function [...]` in the global namespace via some other jupyter cell that outputs javascript did not seem to do the trick either (I think the module codegen explicitly defines the parameter at the beginning).

Curiously, when I use a date dtype (as opposed to datetimes which are what I really want) the formatting works nicely by default.

**Comment by jheer** [comment]
Hmm. Based on your description, my first inclination is that the issue is due to the underlying data type used in DuckDB. If so--and if the timestamps are being represented as type "TIMESTAMP WITH TIME ZONE" in DuckDB--the following PR (which has been merged but not yet released), should fix the issue: https://github.com/uwdata/mosaic/pull/234

**Reply by jheer** [reply]
v0.3.5 is now released with the timestamp update, so I'd recommend giving that a try.

**Reply by frtennis1** [reply]
Thanks for the eyes! v0.3.5 did not do the trick, but I think this was the right idea, just with a type "TIMESTAMP_NS". (I guess pandas timestamps are indeed nanosecond precision, and for some reason this led to it being explicitly in the name?)

Adding that type to the block a-la #234 fixed it for me locally - will try to get a pull-request for it.
